### ISSUE TITLE
Callable with single argument accepts args expanding 

### ISSUE THREAD

**Issue by zhukovgreen** [python/mypy#13992]
**Bug Report**

(A clear and concise description of what the bug is.)

**To Reproduce**

```python
from typing import Callable, TypeVar


T_ = TypeVar("T_")
P_ = TypeVar("P_")


def convert_to_hook(f: Callable[[P_], T_]) -> Callable[[list[P_]], T_]:

    def hook_f(args: list[P_]) -> T_:
        return f(*args)

    return hook_f
```
https://mypy-play.net/?mypy=latest&python=3.10&gist=e45c80cbb1e609c235b4da1813501cd5

**Expected Behavior**

The type check should fail and point to f(*args), because f is accepting only a single argument of type P_.

**Actual Behavior**

Success: no issues found in 1 source file

**Your Environment**

- Mypy version used:
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used:

**Comment by sobolevn** [comment]
I think that the main motivation of why this is not a bug is that we cannot know how many arguments are there in a list.

So, we can either accept all list unpacks or reject all of them.
But, since this is rather popular method of doing things in Python we will generate quite a lot of false positives.

**Comment by zhukovgreen** [comment]
I tried to be more explicit with the number of arguments using Protocol, and it seems the issue is properly captured:
```python
from typing import Callable, TypeVar, Protocol


T_ = TypeVar("T_", covariant=True)
P_ = TypeVar("P_", contravariant=True)

class FuncWithOneArg(Protocol[T_, P_]):
    def __call__(self, arg: P_) -> T_: ...
    
class FuncWithManyArgs(Protocol[T_, P_]):
    def __call__(self, *args: P_) -> T_: ...


def convert_to_hook(f: FuncWithManyArgs) -> FuncWithOneArg:

    def hook_f(args: list[P_]) -> T_:
        return f(*args)

    return hook_f
```

The output is:
```
Failed (exit code: 1) (2811 ms)

main.py:16: error: A function returning TypeVar should receive at least one argument containing the same Typevar
main.py:19: error: Incompatible return value type (got "Callable[[List[P_]], T_]", expected "FuncWithOneArg[Any, Any]")
main.py:19: note: "FuncWithOneArg[Any, Any].__call__" has type "Callable[[Arg(Any, 'arg')], Any]"
Found 2 errors in 1 file (checked 1 source file)
```

See:
https://mypy-play.net/?mypy=latest&python=3.10&gist=550d907fd13458988ad0a7430460ff05
